static/js/formatters-internal.js

```diff
@@ -311,164 +311,117 @@ export function image(simpleOrComplexImage = {}, desiredSize = '200x', atLeastAs
+    let urlPath = url.pathname;
+    if (url.pathname.match('^\/p')) {
+      urlPath = _removePhotoImageUrlExtension(urlPath);
+    }
 
-  const dynamicUrl = isEuImage
-    ? _getEuImageDynamicUrl(image, desiredSize, atLeastAsLarge)
-    : _getUsImageDynamicUrl(image.url, desiredSize, atLeastAsLarge);
+    const hostname = url.hostname.replace(/^[a-z]+\./, 'dyn.');
+    const formatOptionsString = _getImageFormatOptions(desiredSize, atLeastAsLarge, image.width, image.height);
 
-  return Object.assign(
-    {},
-    image,
-    {
-      url: dynamicUrl.replace('http://', 'https://')
-    }
-  );
+    return Object.assign(
+      {},
+      image,
+      {
+        url: `https://${hostname}${urlPath}${formatOptionsString}`,
+      }
+    );
+
+  } catch (error) {
+    throw new Error(`Error processing image url ${image.url}: ${error}`);
+  }
 }
 
 /**
- * Given a US image url, returns the dynamic url.
+ * Construct the format options string with given parameters.
  *
- * @param {string} imageUrl Image's url (e.g.
- *   'https://dynl.mktgcdn.com/p/ldMLwj1JkN94-2pwh6CjR_OMy4KnexHJCfZhPAZCbi0/196x400.jpg',
- *   'https://a.mktgcdn.com/p/ldMLwj1JkN94-2pwh6CjR_OMy4KnexHJCfZhPAZCbi0/196x400.jpg')
  * @param {string} desiredSize The desired size of the image ('<width>x<height>')
  * @param {boolean} atLeastAsLarge Whether the image should be at least as large as the desired
  *                                 size in one dimension or smaller than the desired size in both
  *                                 dimensions.
- * @returns {string} A dynamic url (e.g. 'https://dynl.mktgcdn.com/p/ldMLwj1JkN94-2pwh6CjR_OMy4KnexHJCfZhPAZCbi0/200x1.jpg')
+ * @param {number?} fullSizeWidth The full size width of the original image
+ * @param {number?} fullSizeHeight The full size height of the original image
+ * @returns {string} A string representing the format options (e.g. 'height=200,width=100,fit=cover')
  */
-function _getUsImageDynamicUrl(imageUrl, desiredSize, atLeastAsLarge) {
-  const [urlWithoutExtension, extension] = _splitStringOnIndex(imageUrl, imageUrl.lastIndexOf('.'));
-  const [urlBeforeDimensions, dimensions] = _splitStringOnIndex(urlWithoutExtension, urlWithoutExtension.lastIndexOf('/') + 1);
-  const fullSizeDims = dimensions.split('x');
-
-  let desiredWidth, desiredHeight;
+function _getImageFormatOptions(desiredSize, atLeastAsLarge, fullSizeWidth, fullSizeHeight) {
   let desiredDims = desiredSize.split('x');
+  const desiredWidthProvided = desiredDims[0] !== '' && desiredDims[0] !== '1';
+  const desiredHeightProvided = desiredDims[1] !== '' && desiredDims[1] !== '1';
 
-  if (desiredDims[0] !== '') {
+  // both dimensions are not provided, return original image
+  if (!desiredWidthProvided && !desiredHeightProvided) {
+    return '';
+  }
+
+  const originalRatio =
+    (!!fullSizeWidth && !!fullSizeHeight) ? (fullSizeWidth / fullSizeHeight) : undefined;
+  let desiredWidth;
+  let desiredHeight;
+  let formatOptions = ['fit=contain'];
```

Comment:
don't we want this to be a cover if atLeastAsLarge=true?

Reply:
Per my previous comment, `cover` crops the image rather than respecting the aspect ratio. I do not imagine that is the desired behavior in any case (and it doesn't match the old dynl).

Reply:
By passing `contain` as the formatting option, the image will be shrunk until it fits within the size boundaries. When a user passes in atLeastAsLarge, they are looking for an image size that fully covers the desired dimensions. It's actually a pretty common use case (e.g. for a hero image or banner) to want the image to take up the whole space (or more) rather than be too small and show empty whitespace.

Because the user's desired size may not be the same as the aspect ratio, it's not always possible to both preserve the aspect ratio and keep the whole image while fully covering the area. In these scenarios, we'd return `cover` (because dynl doesn't change the aspect ratio of the image or crop it), and it would be preferable to have an image with the correct aspect ratio that gets cropped than the opposite.

Reply:
I see, that makes sense. Well, as long as the difference in behavior from dynl is understood, that's fine.

Reply:
I think I got mixed up with the EU and US handling and thought we were already doing a cover/crop. It looks like we only do that for EU, and use dynl for US, so one would crop and one wouldn't. To be consistent, we should pick one. I'm leaning towards keeping what we do for US (i.e. don't crop) because that's where most of our current users are. But, it depends on the behavior of dyn. @jfancher-yext, what does dyn return if, via the URL, we pass in:
- a width and a height that don't match the aspect ratio, and no `fit` param
- one dimension (either width or height), and no `fit` param
- no width or height or fit

Reply:
The default `fit` is `scale-down` (not by us, we don't set anything, but it's the apparent default of Cloudflare).

* If width and height are provided, it will choose the largest AR-conforming size that is <= the requested size in both dimensions and <= the base image size. (Or equivalently, the smallest size that is >= the requested size in at least one dimension, but not larger than the base image). Basically the same as dynl except it's continuous instead of a set of fixed sizes.
* If one dimension is given, the largest AR-confirming size that is <= the requested size in that dimension and <= the base image size. Again, like dynl if the other dimension is `1`.
* It is an error to provide no arguments. If you provide other arguments but not width/height/fit then it'll just use the base image size, but if you provide nothing it'll fail at the parsing step. This is probably an oversight, not sure.

Reply:
> - If width and height are provided, it will choose the largest AR-conforming size that is <= the requested size in both dimensions and <= the base image size. (Or equivalently, the smallest size that is >= the requested size in at least one dimension, but not larger than the base image). Basically the same as dynl except it's continuous instead of a set of fixed sizes.

Isn't the first description (largest size smaller than requested) describing dynm and the second (smallest size larger than requested) a description for dynl? And similarly, isn't the second bullet describing dynm, not dynl?

And just to confirm, dyn returns exact pixel match (fully dynamic) rather than sizes chosen from certain thumbnails like dynl and dynm do?

Reply:
Thanks for the clarification! In that case, we can stick with `fit=contain` or switch between `contain` and `cover`. Functionally, they'll be the same. We just have to be careful to always specify the correct dimensions when the user passes in both a desired width and height with atLeastAsLarge=true